src/Disks/ObjectStorages/VFSTransactionLog.cpp

```diff
@@ -0,0 +1,133 @@
+#include "VFSTransactionLog.h"
+#include "Common/ZooKeeper/ZooKeeper.h"
+#include "IO/ReadHelpers.h"
+#include "VFSTraits.h"
+#include "base/EnumReflection.h"
+#include "base/defines.h"
+#include "base/find_symbols.h"
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+}
+
+String VFSTransactionLogItem::serialize() const
+{
+    if (type == Type::CreateInode)
+        return fmt::format("{} {} {} {}", type, remote_path, local_path, bytes_size);
+    return fmt::format("{} {}", type, remote_path);
+}
+
+VFSTransactionLogItem VFSTransactionLogItem::deserialize(std::string_view str)
+{
+    VFSTransactionLogItem out;
+
+    // TODO myrrc proper checking
+    std::vector<std::string_view> parts;
+    splitInto<' '>(parts, str);
+
+    if (parts.size() == 4)
+    {
+        out.type = Type::CreateInode;
+        out.remote_path = parts[1];
+        out.local_path = parts[2];
+        out.bytes_size = parseFromString<size_t>(parts[3]);
+        return out;
+    }
+
+    chassert(parts.size() == 2);
+    out.type = *magic_enum::enum_cast<Type>(parts[0]);
+    out.remote_path = parts[1];
+    return out;
+}
+
+void getStoredObjectsVFSLogOps(
+    VFSTransactionLogItem::Type type, const StoredObjects & objects, Coordination::Requests & ops, const VFSTraits & traits)
+{
+    for (const StoredObject & object : objects)
+    {
+        VFSTransactionLogItem item = static_cast<const VFSTransactionLogItem &>(object);
+        item.type = type;
+        ops.emplace_back(zkutil::makeCreateRequest(traits.log_item, item.serialize(), zkutil::CreateMode::PersistentSequential));
+    }
+}
+
+VFSSnapshot::ObsoleteObjects VFSSnapshot::update(const std::vector<VFSTransactionLogItem> & logs)
+{
+    using enum VFSTransactionLogItem::Type;
+    ObsoleteObjects out;
+
+    // TODO myrrc instead of throwing, we should dedicate a snapshot part to invalid objects
+    // and keep track of them so that some entity could clear object storage
+    for (const VFSTransactionLogItem & item : logs)
+        switch (item.type)
+        {
+            case CreateInode: {
+                if (auto it = items.find(item.remote_path); it != items.end()) [[unlikely]]
+                    throw Exception(
+                        ErrorCodes::LOGICAL_ERROR, "Items with same remote path found in snapshot ({}) and log ({})", it->second.obj, item);
+                items.emplace(item.remote_path, ObjectWithRefcount{item, 0});
+                break;
+            }
+            case Link: {
+                auto it = items.find(item.remote_path);
+                if (it == items.end()) [[unlikely]]
```

Comment:
it seems we don't store currently in the snapshot information about all links to the inode.
It means we can not recover the state of filesystem from the snapshot. 

I think it's better to have not only refcount but also all links. This way we can:
1) reconsile state of filesystem with state in the snapshot
2) wipe out information about dead replica from the snapshot (by unlinking all files with it's prefix)
3) chech refcounts

Reply:
I think that we should not throw but rather dedicate a snapshot part to invalid entries in order to be able to sort them out